Async.Promise: Add "chain" function

lambdalisue opened the pull request:
Vim does not have async/await syntax support thus it's a bit tough work
to write similar function as below (js)

```js
    for (const url of urls) {
        const r = await request(url);
        console.log(r)
    }
```

With this commit, equivalent code of above can be written as

```vim
    let fs = map(urls, { url -> { -> request(url).then({ r -> execute('echo r', '') }) })
    call s:Promise.chain(fs)
```

**Question:** Is there any way to avoid stack-overflow or should I comment on the doc?

rhysd commented on doc/vital/Async/Promise.txt:
```diff
@@ -377,6 +377,40 @@ wait({promise}[, {options}])		*Vital.Async.Promise.wait()*
+	It is like an asynchronous sequential call. It rejects when one of
+	function in {promise_factories} has failed or produced promises
+	rejects. Note that it stop producing promises by functions after
+	rejection.
```
Would you describe what is returned? Documentation only mention the case where it rejects.

rhysd commented on doc/vital/Async/Promise.txt:
```diff
@@ -377,6 +377,40 @@ wait({promise}[, {options}])		*Vital.Async.Promise.wait()*
+	Chain promises produced by {promise_factories} (|List| of |Function|). 
+	It is like an asynchronous sequential call. It rejects when one of
+	function in {promise_factories} has failed or produced promises
+	rejects. Note that it stop producing promises by functions after
```
typo: stop -> stops

lambdalisue commented:
@rhysd Updated. The CI fails but not related to this fix

tyru commented:
@lambdalisue 
Just wondering, I don't want to block this PR.

What is `async for` ? You mean this?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of

Did you intend `do requests in parallel, then output the results in order` ?

tyru commented:
Sorry, doc already mention that. Please ignore.

lambdalisue commented:
My bad. As you said `for await` is correct in JS. The `async for` is Python sorry.

lambdalisue commented:
I've squashed commits and added https://github.com/vim-jp/vital.vim/pull/752/commits/5d00a5c9536ca6c3403190cac3f201282c8d7b96 to test that the number of factories can be beyond the maxfuncdepth.
